### ISSUE TITLE
All CAO models share the same lighting with shaders disabled

### ISSUE THREAD

**Issue by appgurueu** [minetest/minetest#15397]
### Luanti version

Latest master (https://github.com/minetest/minetest/commit/60cd1e452957e2e924eedf9de6b46f6eb7833ccc).

Bisected to https://github.com/minetest/minetest/commit/65af606729f7e3c162bf0b77a02570697f784c66.

### Irrlicht device

SDL

### Operating system and version

Fedora 40

### CPU model

AMD Ryzen 5 3500U

### GPU model

AMD Radeon Vega 11 (integrated graphics)

### Active renderer

OpenGL 4.6.0 (Compatibility Profile), Mesa 24.1.4

### Summary

If you disable shaders, all entities are lit the same. This is particularly noticeable if you play on a crowded server like CTF: Other players will flicker.

I assume this is because of the shared mutable model light values: All entities which share the same model write to those, and whichever CAO happens to write last "wins"; all other ones will be lit wrongly. 

### Steps to reproduce

1. Disable shaders, maybe also set your video driver to `opengl`.
2. Place a maximum strength light source node.
3. Set `/time 0`.
4. Spawn one `testentities:mesh` next to the light source node, another one further afar.

The two meshes should be lit differently: One should be noticeably darker than the other. If they are lit the same, the bug is present. For me this often leads to the mesh which should be lit being darker (assuming it was spawned first):

![Screenshot from 2024-11-07 20-31-08](https://github.com/user-attachments/assets/27827a18-f871-40e7-870f-145416d08615)

but the other way around is also possible (I assume this probably depends on object IDs since we're using a sorted map).

### Fixing this

I believe this problem stems from the fact that all entities share the same meshes from the mesh cache, and hence share the same vertices and colors. Since all of this code runs in a step *before* all the drawcalls are run, only the last set colors are applied. Copying meshes could be considered but I am for multiple reasons, including performance, not a fan of this idea. An alternative would be to set the vertex colors in Irrlicht code, right at draw time.

I would have proposed the following patch to fix this by using the fixed lighting again (what exactly are the platform-specific problems with this?), to be applied directly after the bisected commit:

<details>

```diff
diff --git a/src/client/content_cao.cpp b/src/client/content_cao.cpp
index 5a8e1222f..042f97e4e 100644
--- a/src/client/content_cao.cpp
+++ b/src/client/content_cao.cpp
@@ -757,7 +757,7 @@ void GenericCAO::addToScene(ITextureSource *tsrc, scene::ISceneManager *smgr)
 			m_animated_meshnode->setScale(m_prop.visual_size);
 
 			// set vertex colors to ensure alpha is set
-			setMeshColor(m_animated_meshnode->getMesh(), video::SColor(0xFFFFFFFF));
+			setEmissiveColor(m_animated_meshnode, video::SColor(0xFFFFFFFF));
 
 			setSceneNodeMaterials(m_animated_meshnode);
 
@@ -921,9 +921,9 @@ void GenericCAO::setNodeLight(const video::SColor &light_color)
 		// instead of mutating meshes / buffers for everyone via setMeshColor.
 		// (Note: There are a couple more places here where setMeshColor is used.)
 		if (m_meshnode) {
-			setMeshColor(m_meshnode->getMesh(), light_color);
+			setEmissiveColor(m_meshnode, light_color);
 		} else if (m_animated_meshnode) {
-			setMeshColor(m_animated_meshnode->getMesh(), light_color);
+			setEmissiveColor(m_animated_meshnode, light_color);
 		} else if (m_spritenode) {
 			m_spritenode->setColor(light_color);
 		}
@@ -1376,6 +1376,7 @@ void GenericCAO::updateTextures(std::string mod)
 				material.MaterialType = m_material_type;
 				material.MaterialTypeParam = m_material_type_param;
 				material.TextureLayers[0].Texture = texture;
+				material.Lighting = true;
 				material.BackfaceCulling = m_prop.backface_culling;
 
 				// don't filter low-res textures, makes them look blurry
```

</details>

However it's not as trivial now that the `SMaterial::Lighting` field doesn't exist anymore. We would have to partially undo https://github.com/minetest/minetest/pull/15165.

It would be especially nice if we could get rid of `setMeshColor` here more generally, because then the workaround in https://github.com/minetest/minetest/commit/3064f3ccb769df03ef693bf8ce464481a627512a would become obsolete. It also sucks to be iterating all vertices on the CPU and then having to send that data to the GPU.

**Comment by sfan5** [comment]
Two possibilities:
1. this is an entirely new bug because the commit messed up badly
2. or this is an "old" bug that was also always preset on ogles1/2 with shaders disabled

anyway a simple fix would be to insert a copy mesh call in the right place

> I assume this is because of the shared mutable model light values: All entities write to those, and whichever CAO happens to write last "wins"; all other ones will be lit wrongly.

is it *all* CAOs or just CAOs that share the same model?
the former wouldn't make sense to me.

**Comment by sfan5** [comment]
Alternatively we can of course revert the `Lighting` drop, start using that again and accept the risk of game-breaking regressions for a rendering mode people actually use - days before the release.

**Comment by sfan5** [comment]
>  fixed lighting again (what exactly are the platform-specific problems with this?)

To substantiate this:
It just plainly isn't supported on GLES 2.
And the built-in lighting model is exactly part of the legacy shit that OpenGL 3.2 has pulled the plug on.

> I'm assuming by the rendering mode people actually use you mean the default of shaders enabled?

Yes.

> Why can't we only use Lighting if shaders are disabled?

Effectively that is what we were doing, because shaded materials ignore the lighting model.
This does not change the regression risk.

**Comment by appgurueu** [comment]
> To substantiate this:
> It just plainly isn't supported on GLES 2.

Okay, that is good to know.

> And the built-in lighting model is exactly part of the legacy shit that OpenGL 3.2 has pulled the plug on.

I'm aware, but since this is all legacy code we're talking about here I don't care too much.

Okay, so why not move forward with my 2nd favorite option then: Something along the lines of set mesh colors inside Irrlicht when meshes are about to be drawn based on a hacky material flag for this. I can work on this.

**Comment by sfan5** [comment]
> Something along the lines of set mesh colors inside Irrlicht when meshes are about to be drawn based on a hacky material flag for this. I can work on this.

Sure, if you want to. I think you don't even need a material flag, since the scene node controls drawing on its own.
But ask yourself if this time could not have been better spent.

**Comment by appgurueu** [comment]
> Sure, if you want to.

I don't exactly want to, but I don't see much of an alternative either.

This issue makes playing with shaders disabled very annoying. If you play on CTF with shaders disabled for example, all players constantly flicker. Before 5.10, shaders disabled was the default, so we don't yet have very good information on how much shaders impact e.g. performance, and there are a few data points which would suggest that we should keep fixed pipeline rendering an option for 5.10, which was the plan.

> since the scene node controls drawing on its own

This could also be implemented at a scene node level, but I'm not sure that's right, e.g. because `IMeshSceneNode` and `IAnimatedMeshSceneNode` are two different things (the latter does not extend the former), you might have to duplicate the implementation there or similar.

The way I'm currently seeing it, this is a polyfill compensating for the lack of a generally supported lighting API on the fixed pipeline and thus something like renderer responsibility to be implemented via materials, since that's where `ColorParam` already lives.
If we were to save the old vertex colors this would even be completely opaque to meshes and pretty unlikely to cause any regressions, I think.

The intent behind this flag is that I want to enable this polyfill only for cases where I know it to be necessary (CAOs specifically), otherwise we'd be wastefully (and likely even incorrectly) overwriting vertex colors (e.g. for mapblock meshes).
